### PR TITLE
Add a block whitelist to wilderness explosion regeneration

### PR DESCRIPTION
**Pull request by NigelTicknor**
#### Description: 

This adds a block whitelist to the behavior that regenerates the wilderness after an explosion. Since it is a whitelist, if the config option is empty or not present, all blocks will be considered protected.

I have tested this in every configuration (configured with values, configured empty, and key not present) with the flatfile database, but I have not tested it with the SQL database as I don't have one set up.
____
#### New Nodes/Commands/ConfigOptions: 

Nodes:
wild_revert_on_explosion_block_whitelist: ''
New world settings:
PlotManagementWildRegenBlockWhitelist=

____
#### Relevant Towny Issue ticket:

____
- [x ] I have tested this pull request for defects on a server. 

By making this pull request, I represent that I have the right to waive copyright and related rights to my contribution, and agree that all copyright and related rights in my contributions are waived, and I acknowledge that the TownyAdvanced organization has the copyright to use and modify my contribution under the Towny [License](https://github.com/LlmDl/Towny/blob/master/LICENSE.md) for perpetuity.

### CODE REVIEW

**Comment by LlmDl** [comment]
This looks great, my only issue with it: at first glance it looks like any server that received this update would have no explosions revert until they configured the feature. 

It would be best if there was an option that turned on the whitelist feature (defaults to false in the config.yml,) this way people upgrading don't lose their drag-and-drop update functionality.

I'll end up giving this another read tomorrow after some R&R.

**Comment by NigelTicknor** [comment]
Thanks for taking a look.

A server receiving this update would experience no change to their existing functionality; I just double checked this with my server by deleting the new keys out of `config.yml` and `world.txt` before starting it up.

When the `PlotManagementWildRegenBlockWhitelist` key is not provided (such as when updating from an older version), the whitelist is set to the default value from the node group which is currently empty. Since the whitelist is empty, it is ignored in the call to `isPlotManagementWildRevertProtectingBlock()` meaning the functionality for the server would be the same as in the previous version (all blocks revert). 

If you still want me to add an additional configuration value for enabling the whitelist tomorrow, I would be happy to do that, but I think leaving the whitelist empty to disable the whitelist functionality is reasonbly intuitive.

**Comment by LlmDl** [comment]
> Thanks for taking a look.
> 
> A server receiving this update would experience no change to their existing functionality; I just double checked this with my server by deleting the new keys out of `config.yml` and `world.txt` before starting it up.
> 
> When the `PlotManagementWildRegenBlockWhitelist` key is not provided (such as when updating from an older version), the whitelist is set to the default value from the node group which is currently empty. Since the whitelist is empty, it is ignored in the call to `isPlotManagementWildRevertProtectingBlock()` meaning the functionality for the server would be the same as in the previous version (all blocks revert).
> 
> If you still want me to add an additional configuration value for enabling the whitelist tomorrow, I would be happy to do that, but I think leaving the whitelist empty to disable the whitelist functionality is reasonbly intuitive.

I'll give it a test tomorrow for sure, see what I'm missing.

File: src/com/palmergames/bukkit/towny/object/TownyWorld.java [hunk 1]
```diff
@@ -479,6 +480,33 @@ public boolean isProtectingExplosionEntity(Entity entity) {
...
+
+	public List<String> getPlotManagementWildRevertBlockWhitelist() {
+
+		if (plotManagementWildRevertBlockWhitelist == null)
+			setPlotManagementWildRevertBlockWhitelist(TownySettings.getWildExplosionRevertBlockWhitelist());
+
+		return plotManagementWildRevertBlockWhitelist;
+	}
+
+	public boolean isPlotManagementWildRevertProtectingBlock(Material mat) {
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/object/TownyWorld.java:502]: Change this to isPlotManagementWildRevertWhitelistedBlock() please.

File: src/com/palmergames/bukkit/towny/listeners/TownyBlockListener.java [hunk 2]
```diff
@@ -284,6 +284,8 @@ public void onCreateExplosion(BlockExplodeEvent event) {
 				// Only regenerate in the wilderness.
 				if (!TownyAPI.getInstance().isWilderness(block))
 					continue;
+				if(!townyWorld.isPlotManagementWildRevertProtectingBlock(block.getBlockData().getMaterial()))
+					continue;
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/listeners/TownyBlockListener.java:288]: Remove test here.

File: src/com/palmergames/bukkit/towny/listeners/TownyEntityListener.java [hunk 3]
```diff
@@ -621,6 +621,8 @@ public void onEntityExplode(EntityExplodeEvent event) {
 				// Only regenerate in the wilderness.
 				if (!TownyAPI.getInstance().isWilderness(block))
 					continue;
+				if(!townyWorld.isPlotManagementWildRevertProtectingBlock(block.getBlockData().getMaterial()))
+					continue;
```
Comment (LlmDl) [inline src/com/palmergames/bukkit/towny/listeners/TownyEntityListener.java:625]: Remove test here.

**Review by LlmDl — CHANGES_REQUESTED** [review]
There's a couple of small things, to address (debug message and rename a method.)

I would like you to move move where you're testing if it is a whitelisted block as well. In the TownyRegenAPI class there is a method called isBlacklistedBlock() which you should change to:

```java
	private static boolean isBlacklistedBlock(TownyWorld world, Material type) {
		if (world.getPlotManagementWildRevertBlockWhitelist().isEmpty())
			return world.isPlotManagementIgnoreIds(type);
		else 
			return !world.isPlotManagementWildRevertWhitelistedBlock(type);
	}
```
Which I believe will work to make the non-whitelisted blocks all blacklisted blocks and prevent them starting a regen.

**Comment by NigelTicknor** [comment]
I have implemented the changes you requested.

While I was testing, I noticed some odd behavior that I tracked down to a bug in the current version of Towny: when an explosion event occurs, if the _last_ block in the list was blacklisted, the explosion would be cancelled outright. 

I have made two changes: 
- Don't cancel the explosion when a block is blacklisted (or not whitelisted)
- Don't uncancel the explosion after it is already cancelled

**Comment by LlmDl** [comment]
> I have implemented the changes you requested.
> 
> While I was testing, I noticed some odd behavior that I tracked down to a bug in the current version of Towny: when an explosion event occurs, if the _last_ block in the list was blacklisted, the explosion would be cancelled outright.
> 
> I have made two changes:
> 
> * Don't cancel the explosion when a block is blacklisted (or not whitelisted)
> * Don't uncancel the explosion after it is already cancelled

Indeed, I see it now. I might have to have a bit of a think on this then, currently it's a bit messy looking.

I'm also not certain if the `!event.isCancelled() && ` is needed in 
```java
event.setCancelled(!event.isCancelled() && !TownyRegenAPI.beginProtectionRegenTask(block, count, townyWorld, event));
```
as we are already ignoring any cancelled BlockExplodeEvent.

**Comment by NigelTicknor** [comment]
Sorry if I wasn't clear enough.
The reason I added `!event.isCancelled() &&` is because currently in the `for (Block block : event.blockList())` loop, one block could cancel the event, and the next block could uncancel it, which I assumed is not the intended behavior. It's not to check for cancellations by previous plugins, but rather cancellations by previous iterations of the current loop.

This behavior can be observed in the current Towny version with default config.
Here is an example: If you blow up some terrain in the desert with only sand and one or two gold ore blocks in the explosion radius, the intended result is that the sand regenerates and the ore does not. Depending on the order of the blocks in `event.blockList()` one of two things will happen:
- If a sand block is last, the event is not cancelled, and everything works as expected
- If an ore block is last, the event is cancelled, and no explosion occurs at all

**Comment by LlmDl** [comment]
> Sorry if I wasn't clear enough.
> The reason I added `!event.isCancelled() &&` is because currently in the `for (Block block : event.blockList())` loop, one block could cancel the event, and the next block could uncancel it, which I assumed is not the intended behavior. It's not to check for cancellations by previous plugins, but rather cancellations by previous iterations of the current loop.
> 
> This behavior can be observed in the current Towny version with default config.
> Here is an example: If you blow up some terrain in the desert with only sand and one or two gold ore blocks in the explosion radius, the intended result is that the sand regenerates and the ore does not. Depending on the order of the blocks in `event.blockList()` one of two things will happen:
> 
> * If a sand block is last, the event is not cancelled, and everything works as expected
> * If an ore block is last, the event is cancelled, and no explosion occurs at all

Yeah I'm understanding you, what should happen is the blacklisted block stuff should be getting take of outside of the TownyRegenAPI (probably where you put your initial tests in, after the wilderness test...)

**Comment by NigelTicknor** [comment]
I moved the white/blacklist check out of TownyRegenAPI.

I kept the `!event.isCancelled() &&` check because we still don't want a later check to `!hasProtectionRegenTask(new BlockLocation(block.getLocation()))` to uncancel a previous one.

**Comment by LlmDl** [comment]
Good commit. I want to have a think over it tonight, I may end up accepting this and then splitting apart the hasProtectionRegenTask() from the event.setCancelled() later on myself. Nice first time PR and nice work finding a bug while you're at it. If you use Discord join the Towny discord and I'll get you the plugin-dev role.

**Review by LlmDl — APPROVED** [review]

**Pull request merged**